atecc608a_se.c

```diff
@@ -0,0 +1,226 @@
+    }
+
+    ATCAB_INIT();
+
+    ASSERT_SUCCESS(atcab_read_serial_number(buffer));
+    *buffer_length = ATCA_SERIAL_NUM_SIZE;
+
+exit:
+    ATCAB_DEINIT();
+    return status;
+}
+
+psa_status_t atecc608a_check_config_locked()
+{
+    bool config_locked;
+    psa_status_t status = PSA_ERROR_GENERIC_ERROR;
+    
+    ATCAB_INIT();
+
+    ASSERT_SUCCESS(atcab_is_locked(LOCK_ZONE_CONFIG, &config_locked));
+
+exit:
+    ATCAB_DEINIT();
+    if(status == PSA_SUCCESS)
+    {
+        status = config_locked? PSA_SUCCESS : PSA_ERROR_HARDWARE_FAILURE;
+    }
+    return status;        
+}
+
+psa_status_t atecc608a_export_public_key(psa_key_slot_number_t key,
+                                         uint8_t *p_data, size_t data_size,
+                                         size_t *p_data_length)
+{
+    const size_t key_data_len = 65;
+    const uint16_t slot = key;
+    psa_status_t status = PSA_ERROR_GENERIC_ERROR;
+
+    if (data_size < key_data_len)
+    {
+        return PSA_ERROR_BUFFER_TOO_SMALL;
+    }
+
+    ATCAB_INIT();
+
+    /* atcab_get_pubkey returns concatenated x and y values, and the desired 
+       format is 0x04 + x + y. We start at &p_data[1] and add a 0x04 at p_data[0]. */
+    ASSERT_SUCCESS(atcab_get_pubkey(slot, &p_data[1]));
+
+    p_data[0] = 4;
+    *p_data_length = key_data_len;
+
+#ifdef DEBUG_PRINT
+    printf("atecc608a_export_key - pubkey size %d:\n", *p_data_length);
+    atcab_printbin_sp(p_data, *p_data_length);
+#endif
+
+exit:
+    ATCAB_DEINIT();
+    return status;
+}
+
+psa_status_t atecc608a_asymmetric_sign(psa_key_slot_number_t key_slot,
+                                       psa_algorithm_t alg,
+                                       const uint8_t *p_hash,
+                                       size_t hash_length,
+                                       uint8_t *p_signature,
+                                       size_t signature_size,
+                                       size_t *p_signature_length)
+{
+    const uint16_t key_id = key_slot;
+    psa_status_t status = PSA_ERROR_GENERIC_ERROR;
+
+    /* We can only do ECDSA on SHA-256 */
+    if (alg != PSA_ALG_ECDSA(PSA_ALG_SHA_256) && alg != PSA_ALG_ECDSA_ANY)
+    {
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    if (hash_length != 32)
+    {
+        /* The driver only supports signing things of length 32. */
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    if(signature_size < ATCA_SIG_SIZE)
+    {
+        return PSA_ERROR_BUFFER_TOO_SMALL;
+    }
+
+    ATCAB_INIT();
+
+    /* Signature will be returned here. Format is R and S integers in
+     * big-endian format. 64 bytes for P256 curve. */
+    ASSERT_SUCCESS(atcab_sign(key_id, p_hash, p_signature));
+         
+    *p_signature_length = ATCA_SIG_SIZE;
+
+#ifdef DEBUG_PRINT
+    printf("atecc608a_asymmetric_sign - signature size %d:\n", *p_signature_length);
+    atcab_printbin_sp(p_signature, *p_signature_length);
+#endif
+
+exit:
+    ATCAB_DEINIT();
+    return status;
+}
```

Comment:
To make these functions static, the driver C file should have a global, linker-visible struct filled in with the driver callbacks. The example or anything using the atecc608s can then use the functions through the struct.

Reply:
Sure, we can do it this way if it's the only thing we want to have here. I've even thought about adding a function to print the configuration zone (this might be useful for validation), but after your suggestion - we can have that in an utils.h in the example too.